airbyte-integrations/connectors/source-slack/source_slack/source.py

```diff
@@ -0,0 +1,286 @@
+
+    def request_params(self, **kwargs) -> MutableMapping[str, Any]:
+        p = super().request_params(**kwargs)
+        p["types"] = "public_channel"
+        return p
+
+
+class ChannelMembers(SlackStream):
+    data_field = "members"
+
+    def path(self, **kwargs) -> str:
+        return "conversations.members"
+
+    def request_params(self, stream_state: Mapping[str, Any], batch: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, batch=batch, **kwargs)
+        params["channel"] = batch["channel_id"]
+        return params
+
+    def parse_response(self, response: requests.Response, batch: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping]:
+        for member_id in super().parse_response(response, **kwargs):
+            # Slack just returns raw IDs as a string, so we want to put them in a "join table" format
+            yield {"member_id": member_id, "channel_id": batch["channel_id"]}
+
+    def batches(self, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        channels_stream = Channels(authenticator=self.authenticator)
+        for channel_record in channels_stream.read_records(sync_mode=SyncMode.full_refresh):
+            yield {"channel_id": channel_record['id']}
+
+
+class Users(SlackStream):
+    data_field = "members"
+
+    def path(self, **kwargs) -> str:
+        return "users.list"
+
+
+# Incremental Streams
+def chunk_date_range(start_date: DateTime, interval=1) -> Iterable[Mapping[str, any]]:
+    """
+    Returns a list of the beginning and ending timetsamps of each day between the start date and now.
+    The return value is a list of dicts {'oldest': float, 'latest': float} which can be used directly with the Slack API
+    """
+    intervals = []
+    now = pendulum.now()
+    # Each batch contains the beginning and ending timestamp for a 24 hour period
+    while start_date <= now:
+        end = start_date.add(days=interval)
+        intervals.append({'oldest': start_date.timestamp(), 'latest': end.timestamp()})
+        start_date = start_date.add(days=1)
+
+    return intervals
+
+
+class IncrementalMessageStream(SlackStream, ABC):
+    data_field = "messages"
+    cursor_field = "ts"
+
+    def __init__(self, default_start_date: DateTime, **kwargs):
+        self._default_start_date = default_start_date
+        super().__init__(**kwargs)
+
+    def request_params(self, stream_state: Mapping[str, Any], batch: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, batch=batch, **kwargs)
+        params.update(**batch)
+        return params
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]):
+        current_stream_state = current_stream_state or {}
+        current_stream_state['ts'] = max(float(latest_record['ts']), current_stream_state.get('ts', 0))
```

Comment:
this seems like a pretty common pattern now. not sure if we need to do anything. just noting this.